---
title: "Sarek Paper"
subtitle: "Evaluate Benchmark of HCC1395"
author: "Friederike Hanssen"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    toc: true                               # table of contents
    toc_float: true                        # float the table of contents to the left of the main document content
    toc_depth: 3                            # header levels 1,2,3
    theme: default
    number_sections: true                   # add section numbering to headers
    df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
    css: ./qbic-style.css
    highlight: pygments
    pdf_document: true
---

<!-- QBiC Logo -->

<img src="./QBiCLogo.png" style="position:absolute;top:0px;right:0px;" height="120" width="120"/>

::: watermark
QBiC
:::

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Import all R libraries
library(dplyr)
library(patchwork)
library(forcats)
library(ggplot2)
library(cowplot)
library(knitr)
library(kableExtra)
library(ggpubr)
library(viridisLite)

theme_set(theme_gray())
results_folder <- "results/functional/"
dir.create(results_folder)
```

------------------------------------------------------------------------

# Introduction

For germline benchmarking the datatsets HG002-HG004 were run through the pipeline by combining each mapper with all variant callers. 
For somatic benchmarking a dataset from the HCC1395 cell line was run through the pipeline by combining each mapper with all variant callers. 

# Loading the dataset

## Germline

Truth vcfs and & bed files were downloaded from 

```{r echo=FALSE}
data_30_germline <- read.table(file = "../data/functional/merged_30_germline.csv", sep = ',', header = TRUE) %>%
        mutate('aligner' = case_when(
          grepl("dragmap", run) ~ "Dragmap",
          grepl("bwamem2", run) ~ "BwaMem2",
          grepl("bwa", run) ~ "bwa",
           TRUE ~ "other")) %>%
        mutate('caller' = case_when(
            grepl("strelka", run) ~ "Strelka2_BP",
            grepl("haplotypecaller", run) & !grepl("filtered", run) ~ "HaplotypeCaller",
            grepl("haplotypecaller_filtered", run) ~ "HaplotypeCaller (filtered)",
            grepl("freebayes", run) ~ "FreeBayes",
            grepl("deepvariant", run) ~ "DeepVariant",
             TRUE ~ "other")) %>%
      filter(Filter !='ALL') %>%
      mutate("version" = "sarek3")
```
## Somatic

Truth vcf & bed files were downloaded from [here](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release/v1.2/), following the links from this [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8506910/).

See `script.sh` for details on how `som.py` was run. 

```{r input, echo=FALSE, message=FALSE}
data_30_somatic <- read.table(file = "../data/functional/merged_30_trimmed.csv", sep = ',', header = TRUE) %>%
        mutate('aligner' = case_when(
          grepl("dragmap", run) ~ "Dragmap",
          grepl("bwamem2", run) ~ "BwaMem2",
          grepl("bwa", run) ~ "bwa",
           TRUE ~ "other")) %>%
        mutate('caller' = case_when(
            grepl("strelka", run) ~ "Strelka2_BP",
            grepl("mutect2", run) & !grepl("filtered", run) ~ "Mutect2",
            grepl("mutect2_filtered", run) ~ "Mutect2 (filtered)",
            grepl("freebayes", run) ~ "FreeBayes",
             TRUE ~ "other")) %>%
      filter(type !='records') %>%
      mutate("f_score" = 2 * ((precision*recall)/(precision+recall))) %>%
      mutate("version" = "sarek3")

data_27_somatic <- read.table(file = "../data/functional/merged_27.csv", sep = ',', header = TRUE) %>%
        mutate('aligner' = case_when(
          grepl("bwamem2", run) ~ "BwaMem2",
          grepl("bwa", run) ~ "bwa",
           TRUE ~ "other")) %>%
        mutate('Variant caller' = case_when(
            grepl("strelka_", run)  ~ "Strelka2",
            grepl("strelkabp_", run)  ~ "Strelka2_BP",
            grepl("mutect2", run) & !grepl("filtered", run) ~ "Mutect2",
            grepl("mutect2_filtered", run) ~ "Mutect2 (filtered)",
            grepl("freebayes", run) ~ "FreeBayes",
             TRUE ~ "other")) %>%
      filter(type !='records') %>%
      mutate("f_score" = 2 * ((precision*recall)/(precision+recall)))%>%
      mutate("version" = "sarek2")
```

# Methods

```{r echo = FALSE}

my_palette = palette.colors(palette = "Dark2")

plot_scatter <- function (data, recall, precision, title, outputfile) {
    p <- ggplot(data, aes(!!!ensyms(x=recall, y=precision)), environment = environment()) +
          ggtitle(title) +
          geom_jitter(aes(color=caller, shape = aligner),size = 4, width = 0.001, height = 0.001) +
          theme_pubr() +
          guides(colour = guide_legend(override.aes = list(size=2, alpha = 1))) +
          theme(axis.text.x = element_text(size=12),
               axis.text.y = element_text(size=12),
               axis.title.y = element_text(size=15),
               axis.title.x = element_text(size=15),
               legend.position="top",  legend.box="vertical", legend.margin=margin()) +
          labs(x="Recall", y = "Precision") +
          scale_color_manual(values = my_palette)

    if(!is.na(outputfile)) {
      ggsave(plot=p, filename = paste0(results_folder,outputfile), device="png", width=10, height=10, units="cm")
    }
    return(p)

}

plot_score <- function(data, f1, title, outputfilename) {
  p <- ggbarplot(data, x="aligner", y = f1, fill = "caller", color = "caller", position=position_dodge(width = 0.8), title = title, add = "mean_sd")  +
    theme(axis.text.x = element_text(size=12, angle = 20, hjust=1),
          axis.text.y = element_text(size=12),
          axis.title.y = element_text(size=15),
          axis.title.x = element_blank()) +
    labs(y= "F1-Score") +
    scale_y_continuous(limits=seq(0,1)) +
    scale_fill_manual(values = my_palette) + scale_color_manual(values = my_palette)

  if(!is.na(outputfilename)){
    ggsave(plot=p, filename = paste0(results_folder,outputfilename), device="png", dpi = 600)
  }
  
  return(p)
}

plot_summary <- function(scatter, score, outputfilename){
  
  p <- ggpubr::ggarrange(scatter + labs(shape=NULL, colour=NULL, title=NULL), 
                        score + labs(title=NULL),
                        ncol=2, nrow=1, common.legend = T, legend = "right") 
  ggsave(plot=p, filename = paste0(results_folder, outputfilename), device="png", width = 25, height = 10, units = "cm")
  
  return(p)
}

```

## Germline 

### Overview {.tabset .tabset-fade .tabset-pills}

```{r, echo=FALSE}

data_30_germline_snps = data_30_germline %>% filter(Type !='INDEL')
data_30_germline_indels = data_30_germline %>% filter(Type !='SNP')

data_30_germline_mean <- data_30_germline %>%
                          group_by(aligner, caller, Type) %>%
                          summarise(across(METRIC.Recall, mean, .names = 'mean_recall'),
                                    across(METRIC.Recall, sd, .names = 'sd_recall'),
                                    across(METRIC.Precision, mean, .names = 'mean_precision'),
                                    across(METRIC.Precision, sd, .names = 'sd_precision'),
                                    across(METRIC.F1_Score, mean, .names = 'mean_f1'),
                                    across(METRIC.F1_Score, sd, .names = 'sd_f1'))

data_30_germline_mean_snps <- data_30_germline_mean %>% filter(Type !='INDEL') 
data_30_germline_mean_indels <- data_30_germline_mean %>% filter(Type !='SNP')
```

#### SNVs

```{r, echo=FALSE}
knitr::kable(data_30_germline_mean_snps) %>%
    kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
    scroll_box(width = "100%")
```

#### Indels

```{r, echo=FALSE}
knitr::kable(data_30_germline_mean_indels) %>%
    kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
    scroll_box(width = "100%")
```

### Plots {.tabset .tabset-fade .tabset-pills}

#### SNVs

```{r, echo=FALSE, fig.show="hold", out.width = "50%"}
scatter_snps_germline <- plot_scatter(data_30_germline_mean_snps, "mean_recall", "mean_precision", "Sarek 3.1.1 SNPs Germline", "snps_prec_recall_30_germline.png")
score_snps_germline <- plot_score(data_30_germline_snps , 'METRIC.F1_Score', "Sarek 3.1.1 SNPs Germline", "snps_fscore_311_germline.png")
  
scatter_snps_germline
score_snps_germline

plot_summary(scatter_snps_germline, score_snps_germline, "germline_30_snps_summary.png")
```

#### Indels

```{r echo=FALSE, fig.show="hold", out.width = "50%"}
scatter_indels_germline <- plot_scatter(data_30_germline_mean_indels, 'mean_recall', 'mean_precision', "Sarek 3.1.1 Indels Germline", "indels_prec_recall_30_germline.png")
score_indels_germline <- plot_score(data_30_germline_indels , 'METRIC.F1_Score', "Sarek 3.1.1 Indels Germline", "indels_fscore_311_germline.png")

scatter_indels_germline
score_indels_germline

plot_summary(scatter_indels_germline, score_indels_germline, "germline_30_indels_summary.png")
```


## Somatic {.tabset .tabset-fade .tabset-pills}

```{r echo=FALSE}
data_snvs_30_somatic =  data_30_somatic %>% filter(type !='indels')
data_snvs_27_somatic =  data_27_somatic %>% filter(type !='indels')

data_indels_30_somatic =  data_30_somatic %>% filter(type !='SNVs')
data_indels_27_somatic =  data_27_somatic %>% filter(type !='SNVs')
```
### SNVs {.tabset .tabset-fade .tabset-pills}

#### Recall & Precision


```{r snvs , echo=FALSE, fig.show="hold", out.width = "50%"}
#test <- rbind(data_snvs_30_somatic, data_snvs_2_somatic7) %>% mutate(fill = ifelse(version == "sarek2", caller, NA))
plot_scatter(data_snvs_30_somatic, 'recall', 'precision', "SNVs", "snvs_prec_recall_30_somatic.png")
#plot_scatter(data_snvs_27_somatic, 'recall', 'precision', "Sarek 2.7.2", "snvs_prec_recall_272_somatic.png")
```

#### F_score

```{r , echo=FALSE,  fig.show="hold", out.width = "50%"}

plot_score(data_snvs_30_somatic, 'f_score', "Sarek 3.1.1", "snvs_fscore_311_somatic.png")
#plot_score(data_snvs_27_somatic, 'f_score', "Sarek 2.7.2", "snvs_fscore_272_somatic.png")

```

<!-- ### Indels {.tabset .tabset-fade .tabset-pills} -->

<!-- #### Recall & Precision -->


<!-- ```{r , echo=FALSE, fig.show="hold", out.width = "50%"} -->
<!-- plot_scatter(data_indels_30_somatic, 'recall', 'precision', "Sarek 3.1.1", "indels_pre_recall_311_somatic.png") -->
<!-- plot_scatter(data_indels_27_somatic, 'recall', 'precision', "Sarek 2.7.1", "indels_pre_recall_311_somatic.png") -->
<!-- ``` -->

<!-- #### F_score -->

<!-- ```{r , echo=FALSE,  fig.show="hold", out.width = "50%"} -->

<!-- plot_score(data_indels_30_somatic, 'f_score', "Sarek 3.1.1", "indels_fscore_311_somatic.png") -->
<!-- plot_score(data_indels_27_somatic, 'f_score', "Sarek 2.7.2", "indels_fscore_272_somatic.png") -->

<!-- ``` -->

Package versions:

```{r, echo=F}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
