---
title: "Sarek Paper"
subtitle: "Dataflow investigation"
author: "Friederike Hanssen"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    toc: true                               # table of contents
    toc_float: true                        # float the table of contents to the left of the main document content
    toc_depth: 3                            # header levels 1,2,3
    theme: default
    number_sections: true                   # add section numbering to headers
    df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
    css: ./qbic-style.css
    highlight: pygments
    pdf_document: true
---

<!-- QBiC Logo -->

<img src="./QBiCLogo.png" style="position:absolute;top:0px;right:0px;" height="120" width="120"/>

::: watermark
QBiC
:::

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Import all R libraries
library(dplyr)
library(patchwork)
library(forcats)
library(ggplot2)
library(cowplot)
library(knitr)
library(kableExtra)
library(ggpubr)
library(viridisLite)
library(tidyr)
library(ggpattern)

theme_set(theme_grey())
results_folder <- "results/dataflow"
dir.create(results_folder)

source("./functions.R")
```

------------------------------------------------------------------------

# Introduction

These experiments investigate how splitting affects runtime, and storage usage. CPU & memory were kept at the same values if at all possible

## **Goal**

We added more options to parallelize along the genome.

### Run commands

    nextflow run nf-core/sarek -r 3.1.1 -profile cfc --input --outdir -c trace.config -c custom.config

    custom.config

### For each sample: Split fastq (0,4,8,12,16), default

This runs through mapping, duplicate marking, BQSR, and QC for BWA & non-spark GATK implementation.

|                |           |            |                   |              |                                                                                         |         |                          |
|--------|--------|--------|--------|--------|---------------------|--------|--------|
| sample         | QBiC Code | input size | work folder sizes | trace.config | towerID                                                                                 | resumed |                          |
| HCC1395 normal | none      | 12.5GB     | x                 | x            |                                                                                         | yes     | intervals 251 & fastp 12 |
| HCC1395 tumorl | none      | 14.8       | x                 | x            |                                                                                         | yes     |                          |
| HCC1395 normal | none      | 12.5GB     |                   |              | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/Dev/watch/42RTOMFjQSNjB5> | yes     | intervals 18 & fastp 0   |
| HCC1395 tumorl | none      | 14.8       |                   |              | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/Dev/watch/42RTOMFjQSNjB5> | yes     |                          |
| HCC1395 normal | none      | 12.5GB     |                   |              | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/Dev/watch/2lI1bBtG9kpwDb> | yes     | intervals 126 & fastp 6  |
| HCC1395 tumorl | none      | 14.8       |                   |              | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/Dev/watch/2lI1bBtG9kpwDb> | yes     |                          |
| HCC1395 normal | none      | 12.5GB     |                   |              | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/Dev/watch/4OfvuO8daTSQqi> |         | intervals 500 & fastp 3  |
| HCC1395 tumorl | none      | 14.8       |                   |              | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/Dev/watch/4OfvuO8daTSQqi> |         |                          |

# Loading the dataset

Loading all the individual samples. Print sample summary if possible (e.g. metadata sheet).


```{r echo=FALSE}
execution_fastp12_intervals251 <- read.csv(file = '/Users/monarchy/Projects/sarek_paper/old_stuff/split_intervals/fastp12_intervals251/execution_trace_2022-08-31_17-07-55.txt', sep = "\t", stringsAsFactors = FALSE)

storage_fastp12_intervals251 <- read.csv(file = '/Users/monarchy/Projects/sarek_paper/old_stuff/split_intervals/fastp12_intervals251/folder_sizes.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)

storage_fastp6_intervals125 <- read.csv(file = '/Users/monarchy/Projects/sarek_paper/old_stuff/split_intervals/fastp6_intervals125/folder_sizes.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)

execution_fastp6_intervals125 <- read.csv(file = '/Users/monarchy/Projects/sarek_paper/old_stuff/split_intervals/fastp6_intervals125/execution_trace_2022-09-03_23-47-31.txt', sep = "\t", stringsAsFactors = FALSE)
```

# Methods

<!-- ## Clean data -->

```{r echo=FALSE}

storage_fastp12_intervals251 <- rename_storage_columns(storage_fastp12_intervals251)
storage_fastp6_intervals125 <- rename_storage_columns(storage_fastp6_intervals125)

# Merge both tables on key workdir
merged_fastp12_intervals251 <- merge(execution_fastp12_intervals251, storage_fastp12_intervals251, by = "workdir") %>% 
                                  mutate('fastp' = 12) %>% 
                                  mutate('intervals' = 251)
merged_fastp6_intervals125 <- merge(execution_fastp6_intervals125, storage_fastp6_intervals125, by = "workdir") %>% 
                                  mutate('fastp' = 6) %>% 
                                  mutate('intervals' = 125)

merged <- format_splitting(rbind(merged_fastp12_intervals251, merged_fastp6_intervals125))

```

<!-- ## Function -->

```{r echo=FALSE}
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

plot_time_storage <- function(df_max_time, df, df_storage, group, title, xaxis, outputname, type){
  
    x_axis <- function(type) {
      if(identical('mapping', type)) {
        scale_x_continuous(breaks = c(0, 4, 8, 12, 16), labels = c(0, 4, 8, 12, 16)) }
      else {
        scale_x_continuous(breaks = c(1, 125, 251, 375), labels = c(1, 125, 251, 375))
      }
    }
    
    line <- ggline(df_max_time, x = group, y = "max_y", group=1, palette = cbPalette, color = "sample") +
              x_axis(type) +
              #geom_boxplot(data=df, aes_string(x=group, y="realtime_min", group="sample")) +
              geom_jitter(data=df, aes_string(x=group, y="realtime_min", 
                          group="sample", color="sample"), position=position_dodge(0.7),show.legend = F) +
              labs(y = "Max time/sub-sample (min)", x = xaxis) +
              theme(axis.title.x = element_text(size=15),
                    axis.title.y = element_text(size=15),
                    plot.title = element_text(size = 20, hjust = 0.)) +
              expand_limits(x = 0, y = 0)
    
    bar <- ggbarplot(df_storage, x=group, y="sum", fill = "sample", color = "sample",
                   position = position_dodge(0.7), palette = cbPalette)+
        labs(y = "work dir (GB)", x = xaxis) +
        theme(axis.title.x = element_text(size=15),
              axis.title.y = element_text(size=15),
              plot.title = element_text(size = 20, hjust = 0.),
              legend.text = element_text(size = 13)) +
        rremove("legend.title")

    plot <- ggarrange(line, bar, legend.grob = get_legend(bar))

  ann_plot <-
    annotate_figure(plot, top = text_grob(title,
                face = "bold", size = 14))
  ann_plot
  #ggsave(plot=ann_plot, filename = paste0(results_dir,outputname, ".png"), device="png", dpi = 600
        # width = 20, height = 10, units="cm"
  #     )
  # ggsave(plot=ann_plot, filename = paste0(results_dir,outputname, ".pdf"), device="pdf", 
  #       width=20, height=10, units="cm")
}

plot_splitting_summary <- function(df_time, group, xaxis, type, df_storage, title){
  
    x_axis <- function(type) {
      if(identical('mapping', type)) {
        scale_x_continuous(breaks = c(0, 4, 8, 12, 16), labels = c(0, 4, 8, 12, 16)) }
      else {
        scale_x_continuous(breaks = c(1, 125, 251, 375), labels = c(1, 125, 251, 375))
      }
    }
    
    line <- ggline(df_time, x = group, y = "sum", 
               group="sample",color="sample", palette = cbPalette, numeric.x.axis = FALSE) +
        labs(y = "Time (min)", x = xaxis) +
        theme(axis.title.x = element_text(size=15), 
              axis.title.y = element_text(size=15),
              plot.title = element_text(size = 20, hjust = 0.)) + 
        expand_limits(x = 0, y = 0)  +
        x_axis(type) 

    bar <- ggbarplot(df_storage, x=group, y="sum", 
                 fill = "sample", color = "sample", 
                position = position_dodge(0.7), palette = cbPalette)+
      labs(y = "work dir (GB)", x = xaxis) +
      theme(axis.title.x = element_text(size=15), 
            axis.title.y = element_text(size=15),
            plot.title = element_text(size = 20, hjust = 0.),
            legend.text = element_text(size = 13)) +
      rremove("legend.title")
  
    plot <- ggarrange(line, bar, legend.grob = get_legend(bar)) 
  
    ann_plot <- annotate_figure(plot, top = text_grob(title, face = "bold", size = 14))
    ann_plot
#ggsave(plot=ann_plot, filename = paste0(results_dir,"splitting", ".png"), device="png", dpi = 600)
#     width = 20, height = 10, units="cm")
#ggsave(plot=ann_plot, filename = paste0(results_dir,"splitting", ".pdf"), device="pdf", 
#       width=20, height=10, units="cm")
}
```

## Mapping {.tabset .tabset-fade .tabset-pills}

```{r echo=FALSE, message = FALSE, warning=FALSE}
merge_formatted_mapping <- merged %>% filter(grepl('FASTP|BWAMEM1_MEM|GATK4_MARKDUPLICATES', process))

merged_formatted_fastp <- format_process_mapping(merge_formatted_mapping, 'FASTP')
merged_formatted_bwamem <- format_process_mapping(merge_formatted_mapping, 'BWAMEM1_MEM')
merged_formatted_markdup <- format_process_mapping(merge_formatted_mapping, 'GATK4_MARKDUPLICATES')

### Combine fastp, bwamem, markdup
merged_formatted_mapping_line_time_max <- merge_formatted_mapping %>% 
                                            group_by(fastp, simple_name, sample, status) %>% 
                                            summarise(max_y = max(realtime_min)) %>% 
                                            group_by(fastp, sample, status) %>% 
                                            summarise(sum = sum(max_y))

merged_formatted_mapping_storage_sum <- merge_formatted_mapping %>% 
                                        group_by(fastp, sample, status) %>% 
                                        summarise(sum = sum(workdir_GB))
merged_formatted_mapping_storage_sum$fastp <- as.factor(merged_formatted_mapping_storage_sum$fastp)
```

### FastP

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_fastp$time, 
                  df = merged_formatted_fastp$process,
                  df_storage = merged_formatted_fastp$storage,
                  group = "fastp",
                  title = "FastP",
                  xaxis = "# shards",
                  outputname = "fastp", type = 'mapping')
```

### BWA Mem

```{r echo=FALSE}
  plot_time_storage(df_max_time = merged_formatted_bwamem$time, 
                  df = merged_formatted_bwamem$process,
                  df_storage = merged_formatted_bwamem$storage,
                  group = "fastp",
                  title = "BWAMem",
                  xaxis = "# shards",
                  outputname = "bwamem",
                  type = 'mapping')
```

### Markduplicates

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_markdup$time, 
                  df = merged_formatted_markdup$process,
                  df_storage = merged_formatted_markdup$storage,
                  group = "fastp",
                  title = "GATK4 Markduplicates",
                  xaxis = "# shards",
                  outputname = "markdup",
                  type = 'mapping')
```

### Summary

```{r echo=FALSE}
plot_splitting_summary(df_time = merged_formatted_mapping_line_time_max,
                       group = "fastp",
                       xaxis = "# shards",
                       type = 'mapping',
                       df_storage = merged_formatted_mapping_storage_sum,
                       title = "Sum: FastP, max BWA Mem, Markduplicates")
```

## Base quality score recalibration {.tabset .tabset-fade .tabset-pills}

```{r echo=FALSE, message = FALSE, warning=FALSE}
merge_formatted_bqsr <- merged %>% filter(grepl('BASERECALIBRATOR|GATHERBQSRREPORTS|APPLYBQSR', process))

merged_formatted_baserecalibrator <- format_process_splitting(merge_formatted_bqsr, 'BASERECALIBRATOR')
merged_formatted_gatherbqsr <- format_process_splitting(merge_formatted_bqsr, 'GATHERBQSRREPORTS')
merged_formatted_applybqsr <- format_process_splitting(merge_formatted_bqsr, 'APPLYBQSR')

### Combine BQSR; Gather and apply
merged_formatted_bqsr_line_time_max <- merge_formatted_bqsr %>% 
                                            group_by(intervals, simple_name, sample) %>% 
                                            summarise(max_y = max(realtime_min)) %>% 
                                            group_by(intervals, sample) %>% 
                                            summarise(sum = sum(max_y))

merged_formatted_bqsr_storage_sum <- merge_formatted_bqsr %>% 
                                        group_by(intervals, sample) %>% 
                                        summarise(sum = sum(workdir_GB))
merged_formatted_bqsr_storage_sum$intervals <- as.factor(merged_formatted_bqsr_storage_sum$intervals)
```

### Baserecalibrator

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_baserecalibrator$time, 
                  df = merged_formatted_baserecalibrator$process,
                  df_storage = merged_formatted_baserecalibrator$storage,
                  group = "intervals",
                  title = "GATK4 Baserecalibrator",
                  xaxis = "# interval groups",
                  outputname = "baserecalibrator",
                  type = 'intervals')
```

### GatherBQSRReports

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_gatherbqsr$time, 
                  df = merged_formatted_gatherbqsr$process,
                  df_storage = merged_formatted_gatherbqsr$storage,
                  group = "intervals",
                  title = "GATK4 GatherBQSRReports",
                  xaxis = "# interval groups",
                  outputname = "gatherbqsr",
                  type = 'intervals')
```

### ApplyBQSR

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_applybqsr$time, 
                  df = merged_formatted_applybqsr$process,
                  df_storage = merged_formatted_applybqsr$storage,
                  group = "intervals",
                  title = "GATK4 ApplyBQSR",
                  xaxis = "# interval groups",
                  outputname = "applybqsr",
                  type = 'intervals')
```

### Summary

```{r echo=FALSE}
plot_splitting_summary(df_time = merged_formatted_bqsr_line_time_max,
                       group = "intervals",
                       xaxis = "# interval groups",
                       type = 'intervals',
                       df_storage = merged_formatted_bqsr_storage_sum,
                       title = "Sum: BQSR processes")
```

## Variant calling{.tabset .tabset-fade .tabset-pills}

```{r echo=FALSE, message = FALSE, warning=FALSE}
merge_formatted_variantcalling <- merged %>% filter(grepl('STRELKA_SOMATIC|MUTECT2|FREEBAYES|MERGE_STRELKA|MERGE_MUTECT|MERGE_FREEBAYES', process))

merged_formatted_mutect2 <- format_process_splitting(merge_formatted_variantcalling, '^MUTECT2')
merged_formatted_strelka <- format_process_splitting(merge_formatted_variantcalling, '^STRELKA')
merged_formatted_freebayes <- format_process_splitting(merge_formatted_variantcalling, '^FREEBAYES')
```

### Variant callers (Somatic) {.tabset .tabset-fade .tabset-pills}

#### Mutect2

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_mutect2$time, 
                  df = merged_formatted_mutect2$process,
                  df_storage = merged_formatted_mutect2$storage,
                  group = "intervals",
                  title = "MUTECT2",
                  xaxis = "# interval groups",
                  outputname = "mutec2",
                  type = 'intervals')
```

#### Strelka

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_strelka$time, 
                  df = merged_formatted_strelka$process,
                  df_storage = merged_formatted_strelka$storage,
                  group = "intervals",
                  title = "Strelka",
                  xaxis = "# interval groups",
                  outputname = "strelka",
                  type = 'intervals')
```

#### Freebayes

```{r echo=FALSE}
plot_time_storage(df_max_time = merged_formatted_freebayes$time, 
                  df = merged_formatted_freebayes$process,
                  df_storage = merged_formatted_freebayes$storage,
                  group = "intervals",
                  title = "FreeBayes",
                  xaxis = "# interval groups",
                  outputname = "freebayes",
                  type = 'intervals')
```

#### Summary

Two things: Variantcaller max + merge vcf & all variantcallers

```{r echo = FALSE}
merged_formatted_vc_line_time_max <- merge_formatted_variantcalling %>% 
                                            mutate('simple_name' = case_when(
                                              grepl("MERGE_STRELKA_INDELS", simple_name) ~ "MERGE_STRELKA",
                                              grepl("MERGE_STRELKA_SNVS", simple_name) ~ "MERGE_STRELKA",
                                              TRUE ~ as.character (simple_name)
                                            )) %>%
                                            group_by(intervals, simple_name, sample) %>%
                                            summarise(max_y = max(realtime_min)) %>%
                                            mutate('tool' = case_when(
                                              grepl("STRELKA", simple_name) ~ "strelka",
                                              grepl("MUTECT2", simple_name) ~ "mutect2",
                                              grepl("FREEBAYES", simple_name) ~ "freebayes",
                                              TRUE ~ "other")) %>%
                                            group_by(intervals, tool, sample) %>%
                                            summarise(sum = sum(max_y))

merged_formatted_vc_storage_sum <- merge_formatted_variantcalling %>%
                                        mutate('tool' = case_when(
                                              grepl("STRELKA", simple_name) ~ "strelka",
                                              grepl("MUTECT2", simple_name) ~ "mutect2",
                                              grepl("FREEBAYES", simple_name) ~ "freebayes",
                                              TRUE ~ "other")) %>%
                                        group_by(intervals,tool, sample) %>%
                                        summarise(sum = sum(workdir_GB))
merged_formatted_vc_storage_sum$intervals <- as.factor(merged_formatted_vc_storage_sum$intervals)

x_axis <- function(type) {
      if(identical('mapping', type)) {
        scale_x_continuous(breaks = c(0, 4, 8, 12, 16), labels = c(0, 4, 8, 12, 16)) }
      else {
        scale_x_continuous(breaks = c(1, 125, 251, 375), labels = c(1, 125, 251, 375))
      }
    }
    
    
    
```
```{r}
line <- ggline(merged_formatted_vc_line_time_max, x = "intervals", y = "sum", 
               group="tool",color="tool", palette = cbPalette, numeric.x.axis = FALSE)
    line
    
    bar <- ggbarplot(merged_formatted_vc_storage_sum, x="intervals", y="sum",
                 fill = "tool", color = "tool",
                position = position_dodge(0.7), palette = cbPalette) #+
      # labs(y = "work dir (GB)", x = xaxis) +
      # theme(axis.title.x = element_text(size=15),
      #       axis.title.y = element_text(size=15),
      #       plot.title = element_text(size = 20, hjust = 0.),
      #       legend.text = element_text(size = 13)) +
      # rremove("legend.title")

    plot <- ggarrange(line, bar)

    ann_plot <- annotate_figure(plot, top = text_grob("title", face = "bold", size = 14))
    ann_plot
```
 

# Package versions:

```{r, echo=F}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```