---
title: "Sarek Paper"
subtitle: "Dataflow investigation"
author: "Friederike Hanssen"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    toc: true                               # table of contents
    toc_float: true                        # float the table of contents to the left of the main document content
    toc_depth: 3                            # header levels 1,2,3
    theme: default
    number_sections: true                   # add section numbering to headers
    df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
    css: ./qbic-style.css
    highlight: pygments
    pdf_document: true
---

<!-- QBiC Logo -->

<img src="./QBiCLogo.png" style="position:absolute;top:0px;right:0px;" height="120" width="120"/>

::: watermark
QBiC
:::

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Import all R libraries
library(dplyr)
library(patchwork)
library(forcats)
library(ggplot2)
library(cowplot)
library(knitr)
library(kableExtra)
library(ggpubr)
library(viridisLite)
library(tidyr)
library(ggpattern)

theme_set(theme_grey())
results_folder <- "results/dataflow/"
dir.create(results_folder)

source("./functions.R")
```

------------------------------------------------------------------------

# Introduction

These experiments investigate how splitting affects runtime, and storage usage. CPU & memory were kept at the same values if at all possible

## **Goal**

We added more options to parallelise along the genome.

### Run commands

```         
nextflow run nf-core/sarek -r 3.1.1 -profile cfc --input --outdir -c trace.config -c custom.config --nucleotides_per_second 

custom.config
```

This runs through mapping, duplicate marking, BQSR, and QC, Variant calling for BWA & non-spark GATK implementation.

# Loading the dataset

Loading all the individual samples. Print sample summary if possible (e.g. metadata sheet).

| name               | nucleotides_per_second (number of intervals) | num of cpus for fastp         | tower id                                                                                | work sizes | trace |
|------------|------------|------------|---------------|------------|------------|
| fastp4_intervals78 | 10001 (78)                                   | 4, \--split_fastq 10000000000 | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/cfc/watch/PE2um0F1SrwRi>  | yes        | yes   |
| fastp8_intervals40 | 70000 (40)                                   | 8, \--split_fastq 500000000   | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/cfc/watch/52L0P2tE99JYXs> | yes        | yes   |
| fastp12            | skipped                                      | 12, ---split_fastq 100000000  | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/cfc/watch/ocENvYnNRQAGC>  | yes        | yes   |
| fastp16_intervals1 | 5000000 (1)                                  | 16, \--split_fastq 100000000  | <https://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/cfc/watch/2uPwaXSKrcUaKq> | yes        | yes   |
| fastp0_intervals20 | 200000 (21)                                  | 0, \--split-fastq 0           | <http://cfgateway1.zdv.uni-tuebingen.de/orgs/QBiC/workspaces/cfc/watch/5iFlE6AEOimMO9>  | yes        | yes   |
| intervals10 |  400000 (10)                                  | NA           | https://tower.nf/user/friederike-hanssen/watch/1AmjkxAJTshIcb  | yes        | yes   |

```{r echo=FALSE}

## Fastp 0, intervals 20
storage_fastp0_intervals20 <- read.csv(file = '../data/intervals/fastp0_intervals20/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)
execution_fastp0_intervals20 <- read.csv(file = '../data/intervals/fastp0_intervals20/execution_trace_2023-02-25_22-10-44.txt', sep = "\t", stringsAsFactors = FALSE)

## Fastp 4, intervals 78
storage_fastp4_intervals78 <- read.csv(file = '../data/intervals/fastp4_intervals78/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)
execution_fastp4_intervals78 <- read.csv(file = '../data/intervals/fastp4_intervals78/execution_trace_2023-02-23_12-02-57.txt', sep = "\t", stringsAsFactors = FALSE)

## Fastp 8, intervals 40
storage_fastp8_intervals40 <- read.csv(file = '../data/intervals/fastp8_intervals40/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)
execution_fastp8_intervals40 <- read.csv(file = '../data/intervals/fastp8_intervals40/execution_trace_2023-02-22_15-42-27.txt', sep = "\t", stringsAsFactors = FALSE)

## Fastp 12, intervals 10
#storage_fastp12_intervals10 <- read.csv(file = '../data/intervals/fastp12_intervals10/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)
#execution_fastp12_intervals10 <- read.csv(file = '../data/intervals/fastp12_intervals10/execution_trace_2023-09-19_22-28-30.txt', sep = "\t", stringsAsFactors = FALSE)

## Fastp 12
execution_fastp12_preprocessing <- read.csv(file = '../data/intervals/fastp12/execution_trace_2023-04-21_16-25-01.txt', sep = "\t", stringsAsFactors = FALSE)
storage_fastp12_preprocessing <- read.csv(file = '../data/intervals/fastp12/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)

## intervals 10
execution_intervals10 <- read.csv(file = '../data/intervals/intervals10/execution_trace_2023-11-09_23-33-17.txt', sep = "\t", stringsAsFactors = FALSE)
storage_intervals10 <- read.csv(file = '../data/intervals/intervals10/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)

##  intervals 124
execution_intervals124_preprocessing <- read.csv(file = '../data/bam_vs_cram/cram/cram_nospark_preprocessing_1/execution_trace_2022-11-29_16-01-52.txt', sep = "\t", stringsAsFactors = FALSE) %>%
                                      filter(process != 'NFCORE_SAREK:SAREK:FASTP') %>%
                                      filter(process != "NFCORE_SAREK:SAREK:FASTQ_ALIGN_BWAMEM_MEM2_DRAGMAP:BWAMEM1_MEM") %>%
                                      filter(!grepl("NFCORE_SAREK:SAREK:BAM_MARKDUPLICATES:*", process))

execution_intervals124_variantcalling <- read.csv(file = '../data/bam_vs_cram/cram/cram_variantcalling_1/execution_trace_2023-03-30_10-23-49.txt', sep = "\t", stringsAsFactors = FALSE)

storage_intervals124_preprocessing <- read.csv(file = '../data/bam_vs_cram/cram/cram_nospark_preprocessing_1/folder_sizes.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)
storage_intervals124_variantcalling <- read.csv(file = '../data/bam_vs_cram/cram/cram_variantcalling_1/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)

execution_intervals124 = rbind(#execution_fastp12_preprocessing, 
  execution_intervals124_preprocessing, execution_intervals124_variantcalling)
storage_intervals124 = rbind(#storage_fastp12_preprocessing, 
  storage_intervals124_preprocessing, storage_intervals124_variantcalling)

## Fastp 16, intervals 1
storage_fastp16_intervals1 <- read.csv(file = '../data/intervals/fastp16_intervals0/folder_size.tsv', sep = "\t", stringsAsFactors = FALSE, header = FALSE)
execution_fastp16_intervals1 <- read.csv(file = '../data/intervals/fastp16_intervals0/execution_trace_2023-02-23_18-06-55.txt', sep = "\t", stringsAsFactors = FALSE)


```

# Methods

<!-- ## Clean data -->

```{r echo=FALSE}

storage_fastp0_intervals20 <- rename_storage_columns(storage_fastp0_intervals20)
storage_fastp4_intervals78 <- rename_storage_columns(storage_fastp4_intervals78)
storage_fastp8_intervals40 <- rename_storage_columns(storage_fastp8_intervals40)
storage_intervals124 <- rename_storage_columns(storage_intervals124)
storage_intervals10 <- rename_storage_columns(storage_intervals10)
storage_fastp12_preprocessing <- rename_storage_columns(storage_fastp12_preprocessing)
storage_fastp16_intervals1 <- rename_storage_columns(storage_fastp16_intervals1)

# Merge both tables on key workdir
merged_fastp0_intervals20 <- merge(execution_fastp0_intervals20, storage_fastp0_intervals20, by = "workdir") %>% 
                                  mutate('fastp' = 1) %>% 
                                  mutate('intervals' = 21)
merged_fastp4_intervals78 <- merge(execution_fastp4_intervals78, storage_fastp4_intervals78, by = "workdir") %>% 
                                  mutate('fastp' = 4) %>% 
                                  mutate('intervals' = 78)
merged_fastp8_intervals40 <- merge(execution_fastp8_intervals40, storage_fastp8_intervals40, by = "workdir") %>% 
                                  mutate('fastp' = 8) %>% 
                                  mutate('intervals' = 40)
merged_intervals124 <- merge(execution_intervals124, storage_intervals124, by = "workdir") %>% 
                                  mutate('fastp' = NA) %>% 
                                  mutate('intervals' = 124)
merged_intervals10 <- merge(execution_intervals10, storage_intervals10, by = "workdir") %>% 
                                  mutate('fastp' = NA) %>% 
                                  mutate('intervals' = 10)
merged_fastp12 <- merge(execution_fastp12_preprocessing, storage_fastp12_preprocessing, by = "workdir") %>% 
                                  mutate('fastp' = 12) %>% 
                                  mutate('intervals' = NA)
merged_fastp16_intervals1 <- merge(execution_fastp16_intervals1, storage_fastp16_intervals1, by = "workdir") %>% 
                                  mutate('fastp' = 16) %>% 
                                  mutate('intervals' = 1)


merged <- format_splitting(rbind(merged_fastp0_intervals20, 
                                 merged_fastp4_intervals78, 
                                 merged_fastp8_intervals40,
                                 merged_intervals124,
                                 merged_intervals10,
                                 merged_fastp12,
                                 merged_fastp16_intervals1)) %>% 
  mutate(sample_status = paste0(sample, "_", status))

```

## Mapping {.tabset .tabset-fade .tabset-pills}

```{r echo=FALSE, message = FALSE, warning=FALSE}
merge_formatted_mapping <- merged %>% filter(grepl('FASTP|BWAMEM1_MEM|GATK4_MARKDUPLICATES', process))

merged_formatted_fastp <- format_process_mapping(merge_formatted_mapping, 'FASTP')
merged_formatted_bwamem <- format_process_mapping(merge_formatted_mapping, 'BWAMEM1_MEM')
merged_formatted_markdup <- format_process_mapping(merge_formatted_mapping, 'GATK4_MARKDUPLICATES')

### Combine fastp, bwamem, markdup
merged_formatted_mapping_line_time_max <- merge_formatted_mapping %>% 
                                            group_by(fastp, simple_name, sample_status) %>% 
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>% 
                                            group_by(fastp, sample_status) %>% 
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh'))

merged_formatted_mapping_storage_sum <- merge_formatted_mapping %>% 
                                        group_by(fastp, sample_status) %>% 
                                        summarise(sum = sum(workdir_GB))
merged_formatted_mapping_storage_sum$fastp <- as.factor(merged_formatted_mapping_storage_sum$fastp)
```

### FastP

```{r}
fastp <- plot_dataflow_single_process(df_max_time = merged_formatted_fastp$time, 
                  df = merged_formatted_fastp$process,
                  df_storage = merged_formatted_fastp$storage,
                  group = "fastp",
                  title = "FastP",
                  xaxis = "# shards",
                  outputname = "fastp",
                  results_folder = results_folder)

fastp[[1]]
```

### BWA Mem

```{r echo=FALSE}
bwa <- plot_dataflow_single_process(df_max_time = merged_formatted_bwamem$time, 
                  df = merged_formatted_bwamem$process,
                  df_storage = merged_formatted_bwamem$storage,
                  group = "fastp",
                  title = "BWAMem",
                  xaxis = "# shards",
                  outputname = "bwamem",
                  results_folder = results_folder
                  )
bwa[[1]]
```

### Markduplicates

```{r echo=FALSE}
markdup <- plot_dataflow_single_process(df_max_time = merged_formatted_markdup$time, 
                  df = merged_formatted_markdup$process,
                  df_storage = merged_formatted_markdup$storage,
                  group = "fastp",
                  title = "GATK4 Markduplicates",
                  xaxis = "# shards",
                  outputname = "markdup",
                  results_folder = results_folder
                  )

markdup[[1]]
```

### Combine plots

```{r echo=FALSE}
mapping_arrange <- ggpubr::ggarrange(fastp[[2]], fastp[[3]], fastp[[4]], 
                                     bwa[[2]], bwa[[3]], bwa[[4]],
                                     markdup[[2]],markdup[[3]], markdup[[4]],
                                     ncol=3, nrow=3, legend="none", align="hv", 
                                     labels = c("A", "", "",
                                                "B", "", "",
                                                "C", "", ""))

ggsave(plot=mapping_arrange, filename = paste0(results_folder,"png/supplementary_mapping", ".png"), device="png", width=30, height=30, units="cm")
ggsave(plot=mapping_arrange, filename = paste0(results_folder,"eps/supplementary_mapping", ".eps"), device="eps", width=30, height=30, units="cm")
ggsave(plot=mapping_arrange, filename = paste0(results_folder,"svg/supplementary_mapping", ".svg"), device="svg", width=30, height=30, units="cm")

mapping_arrange
```

### Summary

```{r echo=FALSE}
mapping_summary <- plot_splitting_summary(df_time = merged_formatted_mapping_line_time_max,
                       group = "fastp",
                       xaxis = "# shards",
                       df_storage = merged_formatted_mapping_storage_sum,
                       title = "Sum: Mapping processes",
                       outputname = "summary_mapping")
mapping_summary
```

## Base quality score recalibration {.tabset .tabset-fade .tabset-pills}

```{r echo=FALSE, message = FALSE, warning=FALSE}
merge_formatted_bqsr <- merged %>% filter(grepl('GATK4_BASERECALIBRATOR|GATK4_GATHERBQSRREPORTS|GATK4_APPLYBQSR|MERGE_CRAM', process))

merged_formatted_baserecalibrator <- format_process_splitting(merge_formatted_bqsr, 'GATK4_BASERECALIBRATOR')
merged_formatted_gatherbqsr <- format_process_splitting(merge_formatted_bqsr, 'GATK4_GATHERBQSRREPORTS')
merged_formatted_applybqsr <- format_process_splitting(merge_formatted_bqsr, 'GATK4_APPLYBQSR')
merged_formatted_mergecram <- format_process_splitting(merge_formatted_bqsr, 'MERGE_CRAM')

### Combine BQSR; Gather, apply, and merge
merged_formatted_bqsr_line_time_max <- merge_formatted_bqsr %>%
                                            group_by(intervals, simple_name, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%                                             
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh'))
                                            
merged_formatted_bqsr_storage_sum <- merge_formatted_bqsr %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB))
merged_formatted_bqsr_storage_sum$intervals <- as.factor(merged_formatted_bqsr_storage_sum$intervals)
```

### Baserecalibrator

```{r echo=FALSE}
baserecalibrator <- plot_dataflow_single_process(df_max_time = merged_formatted_baserecalibrator$time,
                  df = merged_formatted_baserecalibrator$process,
                  df_storage = merged_formatted_baserecalibrator$storage,
                  group = "intervals",
                  title = "GATK4 Baserecalibrator",
                  xaxis = "# interval groups",
                  outputname = "baserecalibrator",
                  results_folder = results_folder)

baserecalibrator[[1]]
```

### GatherBQSRReports

```{r echo=FALSE}
gather_bqsr <- plot_dataflow_single_process(df_max_time = merged_formatted_gatherbqsr$time,
                  df = merged_formatted_gatherbqsr$process,
                  df_storage = merged_formatted_gatherbqsr$storage,
                  group = "intervals",
                  title = "GATK4 GatherBQSRReports",
                  xaxis = "# interval groups",
                  outputname = "gatherbqsr",
                  results_folder = results_folder)

gather_bqsr[[1]]
```

### ApplyBQSR

```{r echo=FALSE}
apply_bqsr <- plot_dataflow_single_process(df_max_time = merged_formatted_applybqsr$time,
                  df = merged_formatted_applybqsr$process,
                  df_storage = merged_formatted_applybqsr$storage,
                  group = "intervals",
                  title = "GATK4 ApplyBQSR",
                  xaxis = "# interval groups",
                  outputname = "applybqsr",
                  results_folder = results_folder)

apply_bqsr[[1]]
```

### Merge CRAM

```{r echo=FALSE}
merge_bqsr <- plot_dataflow_single_process(df_max_time = merged_formatted_mergecram$time,
                  df = merged_formatted_mergecram$process,
                  df_storage = merged_formatted_mergecram$storage,
                  group = "intervals",
                  title = "Merge CRAM",
                  xaxis = "# interval groups",
                  outputname = "mergecram",
                  results_folder = results_folder)

merge_bqsr[[1]]
```

### Combine plots

```{r echo=FALSE}

bqsr_arrange <- ggpubr::ggarrange(baserecalibrator[[2]], baserecalibrator[[3]], baserecalibrator[[4]], 
                                  gather_bqsr[[2]], gather_bqsr[[3]], gather_bqsr[[4]], 
                                  apply_bqsr[[2]],apply_bqsr[[3]], apply_bqsr[[4]],
                                  merge_bqsr[[2]], merge_bqsr[[3]], merge_bqsr[[4]],
                                  ncol=3, nrow=4,  
                                  legend="none", align="hv",
                                  labels = c("A", "", "",
                                             "B", "", "",
                                             "C", "", "",
                                             "D", "", ""))
ggsave(plot=bqsr_arrange, filename = paste0(results_folder,"png/supplementary_bqsr", ".png"), device="png", width=30, height=40, units="cm")
ggsave(plot=bqsr_arrange, filename = paste0(results_folder,"eps/supplementary_bqsr", ".eps"), device="eps", width=30, height=40, units="cm")
ggsave(plot=bqsr_arrange, filename = paste0(results_folder,"svg/supplementary_bqsr", ".svg"), device="svg", width=30, height=40, units="cm")

bqsr_arrange
```

### Summary

```{r echo=FALSE}
bqsr_summary <- plot_splitting_summary(df_time = merged_formatted_bqsr_line_time_max,
                       group = "intervals",
                       xaxis = "# interval groups",
                       df_storage = merged_formatted_bqsr_storage_sum,
                       title = "Sum: BQSR processes",
                       outputname = "summary_bqsr")

bqsr_summary
```

## Variant calling {.tabset .tabset-fade .tabset-pills}

### Variant callers (Germline) {.tabset .tabset-fade .tabset-pills}

#### Deepvariant

```{r echo=FALSE}
merged_formatted_deepvariant <- merged %>% filter(grepl('DEEPVARIANT', process))

merged_formatted_deepvariant_time_max <-  merged_formatted_deepvariant %>%
                                            mutate('simple_name_combined' = case_when(
                                                grepl("MERGE_DEEPVARIANT_VCF", simple_name) ~ "MERGE_DEEPVARIANT",
                                                grepl("MERGE_DEEPVARIANT_GVCF", simple_name) ~ "MERGE_DEEPVARIANT",
                                                grepl("TABIX_VC_DEEPVARIANT_VCF", simple_name) ~ "TABIX_DEEPVARIANT",
                                                grepl("TABIX_VC_DEEPVARIANT_GVCF", simple_name) ~ "TABIX_DEEPVARIANT",
                                                TRUE ~ simple_name)) %>%
                                            group_by(intervals, simple_name_combined, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh'))%>%
                                        mutate(caller = 'DeepVariant')

merged_formatted_deepvariant_storage_sum <- merged_formatted_deepvariant %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'DeepVariant')
merged_formatted_deepvariant_storage_sum$intervals <- as.factor(merged_formatted_deepvariant_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_deepvariant_time_max,
                        group = "intervals",
                        xaxis = "# interval groups",
                        df_storage = merged_formatted_deepvariant_storage_sum,
                        title = "Sum: Deepvariant",
                        outputname = "deepvariant")
```

#### Freebayes

```{r echo=FALSE}
merged_formatted_freebayes_germline <- merged %>% filter(grepl('FREEBAYES', process) & status=='normal')

merged_formatted_freebayes_germline_time_max <-  merged_formatted_freebayes_germline %>%
                                            group_by(intervals, simple_name, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh')) %>%
                                            mutate(caller = 'FreeBayes')

merged_formatted_freebayes_germline_storage_sum <- merged_formatted_freebayes_germline %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'FreeBayes')

merged_formatted_freebayes_germline_storage_sum$intervals <- as.factor(merged_formatted_freebayes_germline_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_freebayes_germline_time_max,
                        group = "intervals",
                        xaxis = "# interval groups",
                        df_storage = merged_formatted_freebayes_germline_storage_sum,
                        title = "Sum: FreeBayes (Normal)",
                        outputname = "freebayes_normal")
```

#### Haplotypecaller

```{r echo=FALSE}
merged_formatted_haplotypecaller <- merged %>% filter(grepl('HAPLOTYPECALLER', process))

merged_formatted_haplotypecaller_time_max <-  merged_formatted_haplotypecaller %>%
                                            group_by(intervals, simple_name, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh')) %>%
                                            mutate(caller = 'HaplotypeCaller')

merged_formatted_haplotypecaller_storage_sum <- merged_formatted_haplotypecaller %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'HaplotypeCaller')

merged_formatted_haplotypecaller_storage_sum$intervals <- as.factor(merged_formatted_haplotypecaller_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_haplotypecaller_time_max,
                        group = "intervals",
                        xaxis = "# interval groups",
                        df_storage = merged_formatted_haplotypecaller_storage_sum,
                        title = "Sum: Haplotypecaller",
                        outputname = "haplotypecaller_normal")
```

#### Strelka

```{r echo=FALSE}
merged_formatted_strelka_germline <- merged %>% filter(grepl('STRELKA', process) & status=='normal')

merged_formatted_strelka_germline_time_max <-  merged_formatted_strelka_germline %>%
                                            mutate('simple_name_combined' = case_when(
                                                grepl("MERGE_STRELKA_GENOME", simple_name) ~ "MERGE_STRELKA",
                                                TRUE ~ simple_name)) %>%
                                            group_by(intervals, simple_name_combined, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh'))%>%
                                        mutate(caller = 'Strelka2')

merged_formatted_strelka_germline_storage_sum <- merged_formatted_strelka_germline %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'Strelka2')
merged_formatted_strelka_germline_storage_sum$intervals <- as.factor(merged_formatted_strelka_germline_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_strelka_germline_time_max,
                        group = "intervals",
                        xaxis = "# interval groups",
                        df_storage = merged_formatted_strelka_germline_storage_sum,
                        title = "Sum: Strelka (Normal)",
                        outputname = "strelka_normal")
```

#### Combine Plot

```{r, echo=FALSE}

germline_vc_time <- rbind(merged_formatted_deepvariant_time_max,
                          merged_formatted_freebayes_germline_time_max,
                          merged_formatted_haplotypecaller_time_max,
                          merged_formatted_strelka_germline_time_max)
germline_vc_storage <- rbind(merged_formatted_deepvariant_storage_sum,
                             merged_formatted_freebayes_germline_storage_sum,
                             merged_formatted_haplotypecaller_storage_sum,
                             merged_formatted_strelka_germline_storage_sum)


germline_vc_plot <- plot_variantcaller_summary(df_time = germline_vc_time,
                          group = 'intervals',
                           df_storage = germline_vc_storage,
                           title = 'Germline short variant caller',
                           outputname = 'germline_short_vc_summary')
germline_vc_plot
```

### Variant callers (Somatic) {.tabset .tabset-fade .tabset-pills}

#### Mutect2

```{r echo=FALSE}

merged_formatted_mutect_somatic <- merged %>% filter(grepl('MUTECT2_PAIRED|MERGE_MUTECT2', process) & status == 'paired')

merged_formatted_mutect_time_max <-  merged_formatted_mutect_somatic %>%
                                            group_by(intervals, simple_name, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh')) %>%
                                        mutate(caller = 'Mutect2')

merged_formatted_mutect_storage_sum <-  merged_formatted_mutect_somatic %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'Mutect2')
merged_formatted_mutect_storage_sum$intervals <- as.factor(merged_formatted_mutect_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_mutect_time_max,
                       group = "intervals",
                       xaxis = "# interval groups",
                       df_storage = merged_formatted_mutect_storage_sum,
                       title = "Sum: Mutect2 (Paired)",
                       outputname = "mutect_somatic")
```

#### Strelka

```{r echo=FALSE}
merged_formatted_strelka_somatic <- merged %>% filter(grepl('STRELKA', process) & status=='paired')

merged_formatted_strelka_somatic_time_max <-  merged_formatted_strelka_somatic %>%
                                            mutate('simple_name_combined' = case_when(
                                                grepl("MERGE_STRELKA_INDELS", simple_name) ~ "MERGE_STRELKA",
                                                grepl("MERGE_STRELKA_SNVS", simple_name) ~ "MERGE_STRELKA",
                                                TRUE ~ simple_name)) %>%
                                            group_by(intervals, simple_name_combined, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh'))%>%
                                        mutate(caller = 'Strelka2')

merged_formatted_strelka_somatic_storage_sum <- merged_formatted_strelka_somatic %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'Strelka2')
merged_formatted_strelka_somatic_storage_sum$intervals <- as.factor(merged_formatted_strelka_somatic_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_strelka_somatic_time_max,
                        group = "intervals",
                        xaxis = "# interval groups",
                        df_storage = merged_formatted_strelka_somatic_storage_sum,
                        title = "Sum: Strelka (Paired)",
                        outputname = "strelka_paired")
```

#### Freebayes

```{r echo=FALSE}
merged_formatted_freebayes_somatic <- merged %>% filter(grepl('FREEBAYES', process) & status=='paired')

merged_formatted_freebayes_somatic_time_max <-  merged_formatted_freebayes_somatic %>%
                                            group_by(intervals, simple_name, sample_status) %>%
                                            summarise(across(realtime_min, max, .names = 'max_sample_realtime'),
                                                      across(cpu_h, sum, .names = 'sum_sample_cpuh')) %>%
                                            group_by(intervals, sample_status) %>%
                                            summarise(across(max_sample_realtime, sum, .names = 'sum_combined_per_sample_realtime'),
                                                      across(sum_sample_cpuh, sum, .names = 'sum_combined_per_sample_cpuh')) %>%
                                            mutate(caller = 'FreeBayes')

merged_formatted_freebayes_somatic_storage_sum <- merged_formatted_freebayes_somatic %>%
                                        group_by(intervals, sample_status) %>%
                                        summarise(sum = sum(workdir_GB)) %>%
                                        mutate(caller = 'FreeBayes')

merged_formatted_freebayes_somatic_storage_sum$intervals <- as.factor(merged_formatted_freebayes_somatic_storage_sum$intervals)

plot_splitting_summary(df_time = merged_formatted_freebayes_somatic_time_max,
                        group = "intervals",
                        xaxis = "# interval groups",
                        df_storage = merged_formatted_freebayes_somatic_storage_sum,
                        title = "Sum: FreeBayes (Paired)",
                        outputname = "freebayes_somatic")
```

#### Combined plot

```{r, echo=FALSE}

somatic_vc_time <- rbind(merged_formatted_mutect_time_max,
                         merged_formatted_strelka_somatic_time_max,
                         merged_formatted_freebayes_somatic_time_max)
somatic_vc_storage <- rbind(merged_formatted_mutect_storage_sum,
                            merged_formatted_strelka_somatic_storage_sum,
                            merged_formatted_freebayes_somatic_storage_sum)


somatic_vc_plot <- plot_variantcaller_summary(df_time = somatic_vc_time,
                           group = 'intervals',
                           df_storage = somatic_vc_storage,
                           title = 'Somatic short variant caller',
                           outputname = 'somatic_short_vc_summary')
somatic_vc_plot
```

## Paper plots:

```{r}
paper_plot <- ggpubr::ggarrange(mapping_summary, bqsr_summary,
  common.legend=FALSE, ncol=2, nrow=1, labels = c("A","B"), font.label = list(size = 30))

paper_plot_ann <- annotate_figure(paper_plot, top = text_grob("", face = "bold", size = 14))
 
#vc_suppl_ann_plot
ggsave(plot=paper_plot_ann, filename = paste0(results_folder,"png/paper_dataflow", ".png"), device="png", width=40, height=15, units="cm")
ggsave(plot=paper_plot_ann, filename = paste0(results_folder,"eps/paper_dataflow", ".eps"), device="eps", width=40, height=15, units="cm")
```

# Package versions:

```{r, echo=F}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
