---
title: "Sarek Paper"
subtitle: "Evaluate Benchmark of HCC1395"
author: "Friederike Hanssen"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    toc: true                               # table of contents
    toc_float: true                        # float the table of contents to the left of the main document content
    toc_depth: 3                            # header levels 1,2,3
    theme: default
    number_sections: true                   # add section numbering to headers
    df_print: paged                         # tables are printed as an html table with support for pagination over rows and columns
    css: ./qbic-style.css
    highlight: pygments
    pdf_document: true
---

<!-- QBiC Logo -->

<img src="./QBiCLogo.png" style="position:absolute;top:0px;right:0px;" height="120" width="120"/>

::: watermark
QBiC
:::

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# Import all R libraries
library(dplyr)
library(patchwork)
library(forcats)
library(ggplot2)
library(cowplot)
library(knitr)
library(kableExtra)
library(ggpubr)
library(viridisLite)

theme_set(theme_gray())
results_folder <- "results/functional/"
dir.create(results_folder)
```

------------------------------------------------------------------------

# Introduction

For somatic benchmarking a dataset from the HCC1395 cell line was run through the pipeline by combining each mapper with all variant callers. 

# Loading the dataset

Truth vcf & bed files were downloaded from [here](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release/v1.2/), following the links from this [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8506910/).

See `script.sh` for details on how `som.py` was run. 

```{r input, echo=FALSE, message=FALSE}
data_30 <- read.table(file = "../data/functional/merged_30.csv", sep = ',', header = TRUE) %>%
        mutate('mapper' = case_when(
          grepl("dragmap", run) ~ "Dragmap",
          grepl("bwamem2", run) ~ "BwaMem2",
          grepl("bwa", run) ~ "bwa",
           TRUE ~ "other")) %>%
        mutate('caller' = case_when(
            grepl("strelka", run) ~ "Strelka2_BP",
            grepl("mutect2", run) & !grepl("filtered", run) ~ "Mutect2",
            grepl("mutect2_filtered", run) ~ "Mutect2 (filtered)",
            grepl("freebayes", run) ~ "FreeBayes",
             TRUE ~ "other")) %>%
      filter(type !='records') %>%
      mutate("f_score" = 2 * ((precision*recall)/(precision+recall)))

data_27 <- read.table(file = "../data/functional/merged_27.csv", sep = ',', header = TRUE) %>%
        mutate('mapper' = case_when(
          grepl("bwamem2", run) ~ "BwaMem2",
          grepl("bwa", run) ~ "bwa",
           TRUE ~ "other")) %>%
        mutate('caller' = case_when(
            grepl("strelka_", run)  ~ "Strelka2",
            grepl("strelkabp_", run)  ~ "Strelka2_BP",
            grepl("mutect2", run) & !grepl("filtered", run) ~ "Mutect2",
            grepl("mutect2_filtered", run) ~ "Mutect2 (filtered)",
            grepl("freebayes", run) ~ "FreeBayes",
             TRUE ~ "other")) %>%
      filter(type !='records') %>%
      mutate("f_score" = 2 * ((precision*recall)/(precision+recall)))
```

```{r echo=FALSE, message=FALSE}
data_snvs_30 =  data_30 %>% filter(type !='indels')
data_snvs_27 =  data_27 %>% filter(type !='indels')

data_indels_30 =  data_30 %>% filter(type !='SNVs')
data_indels_27 =  data_27 %>% filter(type !='SNVs')

```

# Methods

```{r}

plot_scatter <- function (data, title, outputfile) {
    p <- ggplot(data, aes(precision,recall, color= caller, fill = caller, shape = mapper)) +
            ggtitle(title) +
            geom_jitter(width= 0.01, height = 0.01) +
            scale_y_continuous(limits=seq(0,1))+
            theme_pubr() + 
            guides(colour = guide_legend(override.aes = list(size=2, alpha = 1))) +
            theme(axis.text.x = element_text(size=15),
                 axis.text.y = element_text(size=15),
                 axis.title.y = element_text(size=15),
                 axis.title.x = element_text(size=15),
                 legend.position="top",  legend.box="vertical", legend.margin=margin())
  
    ggsave(plot=p, filename = paste0(results_folder,outputfile), device="png", dpi = 600)
    print(p)
  
}

plot_score <- function(data, title, outputfilename) {
  p <- ggbarplot(data, x="caller", y = "f_score", fill = "mapper", position=position_dodge(), title = title)  +
    geom_text(aes(label = round(f_score,3), group = mapper), position = position_dodge(.9), vjust = -0.5) +
    theme(axis.text.x = element_text(size=15),
                 axis.text.y = element_text(size=15),
                 axis.title.y = element_text(size=15),
                 axis.title.x = element_text(size=15))
  
  ggsave(plot=p, filename = paste0(results_folder,outputfilename), device="png", dpi = 600)
  print(p)
}


```

## SNVs {.tabset .tabset-fade .tabset-pills}

### Recall & Precision


```{r snvs , echo=FALSE, fig.show="hold", out.width = "50%"}
plot_scatter(data_snvs_30, "Sarek 3.1.1", "snvs_prec_recall_311.png")
plot_scatter(data_snvs_27, "Sarek 2.7.2", "snvs_prec_recall_272.png")
```

### F_score

```{r , echo=FALSE,  fig.show="hold", out.width = "50%"}

plot_score(data_snvs_30, "Sarek 3.1.1", "snvs_fscore_311.png")
plot_score(data_snvs_27, "Sarek 2.7.2", "snvs_fscore_272.png")

```

## Indels {.tabset .tabset-fade .tabset-pills}

### Recall & Precision


```{r , echo=FALSE, fig.show="hold", out.width = "50%"}
plot_scatter(data_indels_30, "Sarek 3.1.1", "indels_pre_recall_311.png")
plot_scatter(data_indels_27, "Sarek 2.7.1", "indels_pre_recall_311.png")
```

### F_score

```{r , echo=FALSE,  fig.show="hold", out.width = "50%"}

plot_score(data_indels_30, "Sarek 3.1.1", "indels_fscore_311.png")
plot_score(data_indels_27, "Sarek 2.7.2", "indels_fscore_272.png")

```

Package versions:

```{r, echo=F}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
